libminifi/src/Connection.cpp

```diff
@@ -268,10 +229,12 @@ void Connection::drain(bool delete_permanently) {
   while (!queue_.empty()) {
     std::shared_ptr<core::FlowFile> item = queue_.front();
     queue_.pop();
-    logger_->log_debug("Delete flow file UUID %s from connection %s", item->getUUIDStr(), name_);
+    logger_->log_debug("Delete flow file UUID %s from connection %s, because it expired", item->getUUIDStr(), name_);
     if (delete_permanently) {
-      if (flow_repository_->Delete(item->getUUIDStr())) {
+      if (item->isStored() && flow_repository_->Delete(item->getUUIDStr())) {
         item->setStoredToRepository(false);
+        auto claim = item->getResourceClaim();
+        if (claim) claim->decreaseFlowFileRecordOwnedCount();
```

Comment:
Shouldn't this call `item->releaseClaim(claim)` instead?

Reply:
the item still owns the `claim_`, and in case somebody holds a reference to it, they should be able to access the content, here we decrement the counter on behalf of the persisted FlowFile instance we just marked for deletion